### ISSUE TITLE
/taskWithTags endpoint ignores errors

### ISSUE THREAD

**Issue by psviderski** [aws/amazon-ecs-agent#2728]
### Summary
The `/v3/taskWithTags` ECS metadata endpoint silently ignores errors when fetching ECS task tags.

### Description
The `/v3/taskWithTags` handler only logs errors when fetching ECS container instance or task tags: https://github.com/aws/amazon-ecs-agent/blob/1489adfaec05c559ea4e1490774d1f86261cf936/agent/handlers/v2/response.go#L146-L172. If there any errors occurred, e.g. `ThrottlingException` communicating with ECS API,  the response will still be successful and may contain an empty or incomplete `TaskTags` map, e.g. it may succeed to get container instance tags but fail to get task tags. It seems there is no way the caller can determine whether all tags are fetched successfully or not. The HTTP status code is always 200.

This issue leads to a Datadog agent collecting tags for containers on an ECS host not being able to retrieve all tags and not retry in case of the rate limiting on the ECS agent end. See the related issue in the Datadog agent: https://github.com/DataDog/datadog-agent/issues/6758

The issue is usually reproducible when deploying a new Datadog agent to our production ECS cluster (50-100 container instances). This results in datadog agents on each host trying to get tags for each container running in the cluster at relatively the same time which required ECS agent to send a lot of requests to AWS ECS API and in turn, results in throttling.

### Expected Behavior
The `/v3/taskWithTags` endpoint which was specifically created for getting task metadata with container instance and task tags retrieved should return an error if it failed to retrieve all tags. For instance, return HTTP status code 429 if `ThrottlingException` occurred.

### Observed Behavior
`/v3/taskWithTags` endpoint returns `200 OK` and an empty or incomplete `TaskTags` map for the task.

### Environment Details
* docker info
```
Client:
 Debug Mode: false

Server:
 Containers: 23
  Running: 23
  Paused: 0
  Stopped: 0
 Images: 11
 Server Version: 19.03.6-ce
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: amazon-ecs-volume-plugin local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ff48f57fc83a8c44cf4ad5d672424a98ba37ded6
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.14.200-155.322.amzn2.x86_64
 Operating System: Amazon Linux 2
 OSType: linux
 Architecture: x86_64
 CPUs: 32
 Total Memory: 123.2GiB
 Name: ip-10-0-10-198.ec2.internal
 ID: YNMS:23UR:WFTC:R5VT:ITOD:L3UO:PUKK:MV2Y:YJBQ:3ZM3:XOJH:XQGU
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```
* curl http://localhost:51678/v1/metadata
```
{
...
Version":"Amazon ECS Agent - v1.47.0 (1489adfa)
}
```

### Supporting Log Snippets

```
2020-11-16-01:7329:level=error time=2020-11-16T01:53:33Z msg="Could not get container instance tags for arn:aws:ecs:us-east-1:***redacted***:container-instance/***redacted***/080c86dd8fc1490c8b748098ee91f7a1: ThrottlingException: Rate exceeded\n\tstatus code: 400, request id: 49ad78db-03b7-4361-bb1f-2c402e34632c" module=response.go
2020-11-16-01:7330:level=error time=2020-11-16T01:53:34Z msg="Could not get container instance tags for arn:aws:ecs:us-east-1:***redacted***:container-instance/***redacted***/080c86dd8fc1490c8b748098ee91f7a1: ThrottlingException: Rate exceeded\n\tstatus code: 400, request id: 09d28b9f-042f-4ce0-a3b3-39d15d4c3f4f" module=response.go
2020-11-16-01:7356:level=error time=2020-11-16T01:53:44Z msg="Could not get container instance tags for arn:aws:ecs:us-east-1:***redacted***:container-instance/***redacted***/080c86dd8fc1490c8b748098ee91f7a1: ThrottlingException: Rate exceeded\n\tstatus code: 400, request id: 0a3f8b43-121f-4c51-b3e9-c2157b2be4ae" module=response.go
2020-11-16-01:7358:level=error time=2020-11-16T01:53:48Z msg="Could not get task tags for arn:aws:ecs:us-east-1:***redacted***:task/***redacted***/2229bf6686204890b8c398719df27faf: ThrottlingException: Rate exceeded\n\tstatus code: 400, request id: ac242366-5931-4c70-8aa0-39c92032e5ac" module=response.go
2020-11-16-01:7359:level=error time=2020-11-16T01:53:49Z msg="Could not get container instance tags for arn:aws:ecs:us-east-1:***redacted***:container-instance/***redacted***/080c86dd8fc1490c8b748098ee91f7a1: ThrottlingException: Rate exceeded\n\tstatus code: 400, request id: c90e9e90-e96f-42c5-89e6-857f10cea115" module=response.go
2020-11-16-01:7392:level=error time=2020-11-16T01:53:55Z msg="Could not get task tags for arn:aws:ecs:us-east-1:***redacted***:task/***redacted***/2229bf6686204890b8c398719df27faf: ThrottlingException: Rate exceeded\n\tstatus code: 400, request id: bc817af1-6973-40ac-87c9-280169501d55" module=response.go
```

**Comment by sharanyad** [comment]
@psviderski Thanks for reporting this issue. Agreed that returning an error response here might make more sense since the path is specifically for task with tags. We will work on changing this.

**Comment by ubhattacharjya** [comment]
Hi @psviderski ,

This https://github.com/aws/amazon-ecs-agent/pull/2789 addresses this problem.

**Comment by ubhattacharjya** [comment]
Hi @psviderski ,

https://github.com/aws/amazon-ecs-agent/pull/2789 address this for v4/taskWithTags and is released with ECS agent version 1.50.0 . Whenever we introduce a change in behavior for task metadata endpoint, we only do it for Task metadata endpoint version 4 as is mentioned in https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v4.html . Kindly use TMDE v4 in order to use this feature.

Closing this issue. Please reopen if you face any issues or need any guidance.

Thanks,
Utsa
